core/src/prediction/CausalPredictionStrategy.cpp

```diff
@@ -0,0 +1,111 @@
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  grf is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with grf. If not, see <http://www.gnu.org/licenses/>.
+ #-------------------------------------------------------------------------------*/
+
+#include <cmath>
+#include <string>
+#include <vector>
+#include <iostream>
+#include "Eigen/Dense"
+#include "commons/utility.h"
+#include "commons/Observations.h"
+#include "prediction/CausalPredictionStrategy.h"
+
+
+const size_t CausalPredictionStrategy::OUTCOME = 0;
+const size_t CausalPredictionStrategy::TREATMENT = 1;
+
+size_t CausalPredictionStrategy::prediction_length() {
+  return 1;
+}
+
+CausalPredictionStrategy::CausalPredictionStrategy(const Data* original_data,
+                                                   const Data* test_data,
+                                                   std::vector<double> lambdas,
+                                                   bool use_unweighted_penalty,
+                                                   std::vector<size_t> linear_correction_variables):
+        original_data(original_data),
+        test_data(test_data),
+        lambdas(lambdas),
+        use_unweighted_penalty(use_unweighted_penalty),
+        linear_correction_variables(linear_correction_variables){
+};
+
+std::vector<double> CausalPredictionStrategy::predict(
+        size_t sampleID,
+        const std::unordered_map<size_t, double>& weights_by_sampleID,
+        const Observations& observations) {
+  size_t num_variables = linear_correction_variables.size();
+  size_t num_total_predictors = original_data->get_num_cols() - 2;
+  size_t num_nonzero_weights = weights_by_sampleID.size();
+
+  std::vector<size_t> indices(num_nonzero_weights);
+  Eigen::MatrixXd weights_vec = Eigen::VectorXd::Zero(num_nonzero_weights);
+  {
+    size_t i = 0;
+    for (auto& it : weights_by_sampleID) {
+      size_t index = it.first;
+      double weight = it.second;
+      indices[i] = index;
+      weights_vec(i) = weight;
+      i++;
+    }
+  }
+
+  Eigen::MatrixXd X (num_nonzero_weights, num_variables + 2);
+  Eigen::MatrixXd Y (num_nonzero_weights, 1);
+  for (size_t i = 0; i < num_nonzero_weights; ++i) {
+    // X columns
+    for (size_t j = 0; j < num_variables; ++j){
+      size_t current_predictor = linear_correction_variables[j];
+      X(i,j+1) = test_data->get(sampleID, current_predictor)
+                 - original_data->get(indices[i],current_predictor);
+    }
+    // treatment W column
+    X(i, num_variables+1) = original_data->get(indices[i], num_total_predictors+1);
```

Comment:
I think it's best to do `observations.get(Observations::TREATMENT, indices[i])`.